Adding dynamic client

krousey opened the pull request:
Attempt to provide a dynamic API client that can access metadata common to API object and lists, and perform common operations on them.

Commit:
Adding dynamic client

krousey commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,184 @@
+	result := new(List)
+	err := rc.namespace(rc.cl.Get()).
+		Resource(rc.resource.Name).
+		VersionedParams(&opts, api.Scheme). // What to do about VersionedParams converter?
```
Is it ok that we use api.Scheme for conversion here? I think the whole idea behind a dynamic client is not needing to know anything (like how to convert) about the underlying structure.

lavalamp replied:
Yeah, I think it's appropriate to take the standard ListOptions here. But maybe from the v1 package and not the unversioned api package?

smarterclayton replied:
This should be using a generic parametercodec created specifically for the "common API" things.  It also assumes that the endpoint supports this version of list params - ultimately that means that the client has a certain "level" of list options that the server may not match.

krousey replied:
Do you mean `api.ParameterCodec`? I've run into a problem with that. It fails to recognize the `ListOptions` belonging to an arbitrary `GroupVersion` if the `GroupVersion` isn't registered. This means the conversion step fails and results in an error.

I was thinking of just writing a parameter codec that doesn't do group conversion and just calls `queryparams.Convert`. Is this a bad idea?

smarterclayton replied:
I think there is a generic.ParameterCodec that only needs to know about all of the possible list options a server could support.  To make progress you could have a big TODO in there "deal with conversion".  I think dealing with conversion will require the discovery doc to describe what the common schema is so you can prime generic.ParameterCodec.  For now "best effort' is fine.

krousey replied:
I'm having trouble finding the ParameterCodec you mention. The only `EncodeParameters` implementation is on `runtime.parameterCodec` and the only call to `runtime.NewParameterCodec` is in pkg/api/register.

smarterclayton replied:
There is no generic codec yet - just giving it that name for convenience
here.  dynamic.ParameterCodec

On Wed, Feb 3, 2016 at 1:33 PM, krousey <notifications@github.com> wrote:

> In pkg/client/typed/dynamic/client.go
> <https://github.com/kubernetes/kubernetes/pull/20351#discussion_r51765129>
> :
>
> > +// namespace applies a namespace to the request if the configured
> > +// resource is a namespaced resource. Otherwise, it just returns the
> > +// passed in request.
> > +func (rc *ResourceClient) namespace(req *client.Request) *client.Request {
> > +	if rc.resource.Namespaced {
> > +		return req.Namespace(rc.ns)
> > +	}
> > +	return req
> > +}
> > +
> > +// List returns a list of objects for this resource.
> > +func (rc *ResourceClient) List(opts api.ListOptions) (*List, error) {
> > +	result := new(List)
> > +	err := rc.namespace(rc.cl.Get()).
> > +		Resource(rc.resource.Name).
> > +		VersionedParams(&opts, api.Scheme). // What to do about VersionedParams converter?
>
> I'm having trouble finding the ParameterCodec you mention. The only
> EncodeParameters implementation is on runtime.parameterCodec and the only
> call to runtime.NewParameterCodec is in pkg/api/register.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351/files#r51765129>.
>

krousey commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,184 @@
+// Object is an arbitrary resource with metadata exposed.
+type Object struct {
+	unversioned.TypeMeta `json:",inline"`
+	api.ObjectMeta       `json:"metadata,omitempty"` // There's probably something better for this...
```
Is there something analogous to ListMeta for general objects? Is this it? This data was the one reason I didn't re-use Unstructured. I needed to be about to pull out names for Update above. If this type of data isn't needed, I could re-use Unstructured. Though Unstructured does decode the entire message into a map[string]interface{} where this leaves it as raw bytes.

lavalamp replied:
Yes, this is "ListMeta for non-lists" ;)

(Consider using it from the versioned package?)

We have runtime.Unstructured and runtime.Unknown; my preference would be to reuse one of those objects and change it as necessary, it should be OK to add metadata to one of those.

May need to split into UnstructuredObject and UnstructuredList, now that I think about it...

smarterclayton replied:
Yeah, UnstructuredObject was not really "Objects", it was anything.  Having a concrete Object and List split would be good.

smarterclayton replied:
ObjectMeta in particular is a type that has evolved over time as we add stuff.  I'd expect a few more fields to be added in a backcompat way.

smarterclayton replied:
I think this can be replaced totally by something that can look at unstructured (map[string]interface{}) and attempt to find a matching schema like api/meta does for structs.

krousey commented:
This is still a work in progress. I have more tests to write, as I'm not that confident in Watch yet.

There's a couple of open questions I left as comments in the code. I'll go back and comment on them directly. There's also a question about organization. The new types and codec are in this package. Is this OK? Is there a better place for them to live? It didn't make sense to register the codec, since it's not really associated to any version.

Open to all comments about how to do this better.

cc @lavalamp 

cc @derekwaynecarr since you expressed interested in the SIG meeting
cc @smarterclayton To spot check how I implemented the codec considering that I don't want it to do any conversion. Did I appropriately ignore defaults and overrides?

krousey commented:
@kubernetes/sig-api-machinery

lavalamp commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,184 @@
+// codec is ignored, as the dynamic client uses it's own codec.
+func NewClient(conf *client.Config) (*Client, error) {
+	// avoid changing the original config
+	confCopy := *conf
```
Shallow copy is OK?

krousey replied:
Precedent: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/discovery_client.go#L220

smarterclayton replied:
We should probably be careful on reviews of client config to ensure that holds true.  Probably worth it to add a copy the next time we change client.config.

lavalamp commented:
Looks good so far!

smarterclayton commented on pkg/client/typed/dynamic/codec.go:
```diff
@@ -0,0 +1,159 @@
+
+// JSONCodec is a runtime.Codec that attempts to dynamically decode
+// objects and lists metadata for generic consumption.
+type JSONCodec struct {
```
I think this is what UnstructuredCodec was supposed to do

krousey replied:
Should I start working on `UnstructuredObject`, `UnstructuredList`, and moving some of this logic back into the `UnstructuredJSONScheme`? Or do you think this should remain separate?

lavalamp replied:
The former sounds like a good idea to me.

lavalamp replied:
Also I guess now that @smarterclayton's rename has gone in, this should be called a "Serializer" not a "Codec".

smarterclayton replied:
I would roll it into to UnstructuredJSONScheme as well - agree this is a Serializer (does not guarantee versioning).

krousey commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,194 @@
+
+var pCodec runtime.ParameterCodec = parameterCodec{}
+
+func getObjectName(obj *runtime.Unstructured) (string, error) {
```
This is what I used instead of including metadata in the unstructured objects.

smarterclayton replied:
Leave a TODO that this needs to be centralized

smarterclayton replied:
On Status in decode, I'm surprised you need to do that - request.go
should call decodeInto on your codec to check for status.

Having a wrapper that lives in typed client and uses the returned GVK
to transform specific kinds into known API types would probably be the
short term recommendation.  Another alternative is to create a scheme
and import specifically unversioned status and the other common types
(we need this anyway) and do a delegation call into the scheme.

krousey commented:
@lavalamp @smarterclayton 

I think I addressed the high level comments you both had. If you guys don't have any more high level ones, I'm going to add more tests and turn this in to a real review.

smarterclayton commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,194 @@
+	return errors.New("DecodeParameters not implemented on nilParameterCodec")
+}
+
+var pCodec runtime.ParameterCodec = parameterCodec{}
```
we generally don't use this naming style - it should be `parameterCodec`

smarterclayton commented on pkg/runtime/unstructured.go:
```diff
@@ -30,13 +33,96 @@ var UnstructuredJSONScheme Codec = unstructuredJSONScheme{}
+		return nil, err
+	}
+
+	if strings.ToLower(det.Kind) == "status" {
```
I don't think this belongs in this method, or in this package

smarterclayton replied:
If anything, you should be doing a second decode of unversioned.Status in the client library in the event you don't receive the data, or at minimum wrap the codec you pass into that type.

krousey replied:
Yea this felt weird. I think I will wrap the codec for this.

smarterclayton commented on pkg/runtime/unstructured.go:
```diff
@@ -30,13 +33,96 @@ var UnstructuredJSONScheme Codec = unstructuredJSONScheme{}
+		return nil, nil, err
+	}
+
+	gvk := obj.GetObjectKind().GroupVersionKind()
```
You should be using the "kind" and "apiVersion" fields out of the decoded object if you didn't call decodeInto

krousey replied:
I am. s.decode assigns in to obj.

derekwaynecarr commented:
Can you add DeleteCollection?

On Friday, January 29, 2016, krousey <notifications@github.com> wrote:

> Attempt to provide a dynamic API client that can access metadata common to
> API object and lists, and perform common operations on them.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/kubernetes/kubernetes/pull/20351
> Commit Summary
>
>    - WIP Adding dynamic client
>
> File Changes
>
>    - *A* pkg/client/typed/dynamic/client.go
>    <https://github.com/kubernetes/kubernetes/pull/20351/files#diff-0>
>    (184)
>    - *A* pkg/client/typed/dynamic/client_test.go
>    <https://github.com/kubernetes/kubernetes/pull/20351/files#diff-1>
>    (108)
>    - *A* pkg/client/typed/dynamic/codec.go
>    <https://github.com/kubernetes/kubernetes/pull/20351/files#diff-2>
>    (159)
>    - *A* pkg/client/typed/dynamic/codec_test.go
>    <https://github.com/kubernetes/kubernetes/pull/20351/files#diff-3>
>    (94)
>
> Patch Links:
>
>    - https://github.com/kubernetes/kubernetes/pull/20351.patch
>    - https://github.com/kubernetes/kubernetes/pull/20351.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351>.
>

krousey commented:
@derekwaynecarr From what I can tell, only events support DeleteCollection. What is the expected behavior if this were used with Pods or Services? I was trying to limit this to the common operations.

derekwaynecarr commented:
DeleteCollection should work for all namespaced content except Services,
but that needs to be fixed.

See: https://github.com/kubernetes/kubernetes/issues/20581

On Tuesday, February 9, 2016, krousey <notifications@github.com> wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> From what I can tell,
> only events support DeleteCollection. What is the expected behavior if this
> were used with Pods or Services? I was trying to limit this to the common
> operations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351#issuecomment-182126015>
> .
>

derekwaynecarr commented:
If it's not supported, it's fine to return the error that comes back from
the API server.  So for services, I would expect an error response, but for
all others, it should be fine.

On Tuesday, February 9, 2016, krousey <notifications@github.com> wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> From what I can tell,
> only events support DeleteCollection. What is the expected behavior if this
> were used with Pods or Services? I was trying to limit this to the common
> operations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351#issuecomment-182126015>
> .
>

krousey commented:
@derekwaynecarr I only see a client side implementation on events. Can I assume this to be a canonical was to call with with respect to query parameters, body, and HTTP Verb? That is a DELETE verb, list options in the query parameter, delete options in the body?

derekwaynecarr commented:
All of the generated clients support the call.

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/typed/generated/core/unversioned/pod.go#L36

Every registry that uses etcd generic for storage has the server side
support:

https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/generic/etcd/etcd.go#L444

Unfortunately services does not use that registry yet, but all others
should be good to go.



On Tuesday, February 9, 2016, krousey <notifications@github.com> wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> I only see a client
> side implementation on events. Can I assume this to be a canonical was to
> call with with respect to query parameters, body, and HTTP Verb? That is a
> DELETE verb, list options in the query parameter, delete options in the
> body?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351#issuecomment-182135989>
> .
>

krousey commented:
OK I think this is ready for full review. I picked @lavalamp, but feel free to reassign. @smarterclayton @derekwaynecarr feel free to keep commenting :smile:

cc @kubernetes/sig-api-machinery @kubernetes/goog-csi

derekwaynecarr commented:
Is this making kube 1.2?  I am excited to use this...

lavalamp commented:
@derekwaynecarr me too-- you can increase the chances by helping me review it? ;)

derekwaynecarr commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,235 @@
+	conf.Codec = dynamicCodec{}
+
+	if conf.APIPath == "" {
+		conf.APIPath = "/api"
```
I thought the unversioned client had constants for this...  grep legacyAPIPath, not a big deal though

krousey replied:
It does. It's not public. I'm not sure I'd want that public anyways. Would you like it to be a constant instead of a literal?

derekwaynecarr replied:
Is it possible to refactor https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/helper.go#L382 to basically not do the codec piece... mostly, I worry about drift on the default qps and burst values.

krousey replied:
It's not the codec that is the problem. It only sets it if it's non-nil. It's the checking for the registered group. The whole point of this client is that the group doesn't have to be registered.

derekwaynecarr replied:
ok

krousey replied:
That being said, I wouldn't mind if you filed an issue to unify some of this later and assign that to me.

derekwaynecarr commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,235 @@
+// Resource returns an API interface to the specified resource for
+// this client's group and version. If resource is not a namespaced
+// resource, then namespace is ignored.
+func (c *Client) Resource(resource *unversioned.APIResource, namespace string) *ResourceClient {
```
I kind of expected to see an interface for what ResourceClient supports to help mock use cases...  will have to look at what you have in rest of PR but recording it here as I read through...

krousey replied:
This was on purpose. Ideally, you shouldn't need a mock. You should be able to have this talk to a fake server, like I did in the unit tests.

If this had been an interface, and checked in, and someone mocked it for a unit test, we wouldn't be able to change it without breaking their tests. For example, I would not have been able to add DeleteCollections without adding a DeleteCollections to their mock.

I still suspect that everyone will ask for an interface anyways.

derekwaynecarr replied:
A minor nit, I am not sure its worth changing, but `unversioned.APIResource` has the resource, if its namespaced, and the kind... in practice, you ignore kind, it would be cool if we had an object that excluded kind that was passed in here...

krousey replied:
I chose this to match what was returned by the Discovery client. I think in practice, it will be less annoying to just reuse what Discovery client gives us.

derekwaynecarr commented:
I would really like a fake client, similar to how `fake.NewSimpleClientset` works that lets me get a list of `mockClient.Actions()` that it has recorded.... to do that, I think the `ResourceClient` should implement an interface of some kind so I can swap out in testing.

krousey commented:
@derekwaynecarr Can I follow up with a testing package for this? If I can't support something like what you want without an interface, can I change it then?

derekwaynecarr commented:
@krousey - I think we can follow-up on the testing and interface needs.  I ask because when I convert the namespace controller to use this client with the discovery client, I will want the test client to test the actions it invoked.

derekwaynecarr commented:
One last request, then this is LGTM, but can you add an integration test that exercises the client here:

https://github.com/kubernetes/kubernetes/blob/master/test/integration/client_test.go

smarterclayton commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,235 @@
+}
+
+func (parameterCodec) DecodeParameters(parameters url.Values, from unversioned.GroupVersion, into runtime.Object) error {
+	return errors.New("DecodeParameters not implemented on nilParameterCodec")
```
typo?

krousey replied:
Changed name without changing the error. Will fix with the integration test.

smarterclayton commented:
No more comments from me, previous changes like fine

krousey commented:
@derekwaynecarr I'm not that familiar with that test package. Would you want a version of all 3 of those tests, but using the dynamic client where it makes sense?

derekwaynecarr commented:
@krousey - I am looking for an integration test that could just do the following:

1. use a typed client to create a pod
2. use the dynamic client to list pod
3. use the dynamic client to get pod
4. use the dynamic client to delete pod
5. verify the pod is actually deleted as expected using the typed client

krousey commented:
@derekwaynecarr Awesome. I will do that. 

As an aside, should I be worried that `//+build integration,!no-etcd` is not a valid build tag? It has to be alpha-numeric. Do we even use `no-etcd`?

krousey commented:
@derekwaynecarr I added the test.

@smarterclayton I went ahead and addressed the TODO you had me put in. I want to make sure how I exposed name is ok by you.

derekwaynecarr commented:
Test looks good , thanks!  Will let Clayton review the todo and we can then
merge.

On Thursday, February 11, 2016, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e test build/test *passed* for commit 542d673
> <https://github.com/kubernetes/kubernetes/commit/542d673cda786a76e2b10e221d905025aa0463cd>
> .
>
>    - Build Log
>    <http://pr-test.k8s.io/20351/kubernetes-pull-build-test-e2e-gce/28177/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/20351/kubernetes-pull-build-test-e2e-gce/28177/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//28177>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351#issuecomment-183138374>
> .
>

krousey commented:
@smarterclayton @lavalamp I think we're just waiting on an explicit ACK to the unstructured change I made to expose names in https://github.com/krousey/kubernetes/commit/542d673cda786a76e2b10e221d905025aa0463cd before we squash and LGTM this.

smarterclayton commented:
LGTM

smarterclayton commented:
Squash and self merge

krousey commented:
Squashed and applying LGTM per @smarterclayton and @derekwaynecarr

caesarxuchao commented on pkg/client/typed/dynamic/client.go:
```diff
@@ -0,0 +1,217 @@
+		return nil, nil, err
+	}
+
+	if _, ok := obj.(*unversioned.Status); !ok && strings.ToLower(gvk.Kind) == "status" {
```
Why is it `!ok`? I thought it should be `ok`.

caesarxuchao replied:
Oh, I got it.

krousey commented:
@k8s-bot e2e test this

derekwaynecarr commented:
ok to test

krousey commented:
@k8s-merge-robot e2e test this

derekwaynecarr commented:
I tried to use this and ran into some problems.  Can we defer creation of the RESTClient until after I specify a resource?  Otherwise, this is really hard to consume in our controllers?  Basically, hold a copy of the kubeConfig given at NewClient and then copy and set the group version after I specify a resource?

krousey commented:
@derekwaynecarr It's possible, but I would like to understand why this is a problem. How does deferring the creation of the rest client help if the API the client produces is exactly the same?

derekwaynecarr commented:
@krousey - I will push an example for the namespace controller using this for where it causes the problem.

derekwaynecarr commented:
@krousey - here is a PR where I tried to start using this:

I want to be able to construct a dynamic client in the controllermanager.go here:
https://github.com/kubernetes/kubernetes/commit/0fab678ec5a3793fde6c69b00dac515ea0dbc6cf#diff-2aa9cfe6a01016c1679752b0fcfc9244R248

The issue is that at this point, it has no notion of the group version to use.

Later, I do some discovery, and know the set of resources I plan to delete, and will call delete collection here:
https://github.com/kubernetes/kubernetes/commit/0fab678ec5a3793fde6c69b00dac515ea0dbc6cf#diff-f86a8dd1c06d1f9578423999f8094cc7R138

At this point, when I actually know api resource is when I want the REST client built, not before discovery ever happened.

Make sense?

derekwaynecarr commented:
I want to avoid passing a *client.Config into the namespace controller as it makes testing the controller really difficult versus passing in a mock client implementation.

derekwaynecarr commented:
After discussion on Slack, we agreed to put an interface in front of `dynamic.Client` in a separate PR that will maintain a cache of clients per group version.  The namespace controller will use that.

krousey commented:
@k8s-bot e2e test this issue: #21257

krmayankk commented:
What does this client provide ? Is this different from unversion go client for kubernetes or  is it a layer on top of it ? Does it remove the 70+ dependencies the earlier go client depends on ?

krousey commented:
@k8s-bot e2e test this issue: #21257

krmayankk commented:
@derekwaynecarr What does this client provide ? Is this different from unversioned go client for kubernetes or is it a layer on top of it ? Does it remove the 70+ dependencies the earlier go client depends on ?

krousey commented:
@krmayankk This client provides the ability to minimally interact with resources that weren't necessarily compiled in. The namespace controller (something @derekwaynecarr works on) has a use case to delete all resources under a namespace when the namespace is being deleted. This gives him a generic way to do that.

Another use case is to allow eventually allow kubectl to interact with 3rd party API extensions without having to compile them in.

Regarding dependencies, no. It might have few dependencies, but that was not an explicit goal for this PR. Also, this client isn't as fully featured as the others.


There are other efforts to reduce the dependencies (API restructuring, client generation...), but this is not one of them.

krousey commented:
@k8s-bot e2e test this issue: #21138

lavalamp commented:
@krmayankk we will work on reducing the import tree next. First we're
trying to get the client into the right shape at all...

On Thu, Feb 18, 2016 at 10:05 AM, krousey <notifications@github.com> wrote:

> @k8s-bot <https://github.com/k8s-bot> e2e test this issue: #21138
> <https://github.com/kubernetes/kubernetes/issues/21138>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20351#issuecomment-185838263>
> .
>

krousey commented:
@k8s-bot e2e test this issue: #21463

krousey commented:
I'm adding the e2e-not-required label since e2e is flaky and this code in no way affects the e2e suite.
